src/backends/libtorch_c/torch_extensions/torch_redis.cpp

```diff
@@ -16,54 +18,125 @@ torch::IValue IValueFromRedisReply(RedisModuleCallReply *reply){
+            throw(std::runtime_error("Unexpected internal error"));
         }
     }
 }
 
-torch::IValue redisExecute(std::string fn_name, std::vector<std::string> args ) {
-    RedisModuleCtx* ctx = RedisModule_GetThreadSafeContext(NULL);
+torch::IValue redisExecute(const std::string& fn_name, const std::vector<std::string> &args ) {
+    RedisModuleCtx* ctx = RedisModule_GetThreadSafeContext(nullptr);
     RedisModule_ThreadSafeContextLock(ctx);
     size_t len = args.size();
     RedisModuleString* arguments[len];
     len = 0;
-    for (std::vector<std::string>::iterator it = args.begin(); it != args.end(); it++) {
-        const std::string arg = *it;
+    for (auto &arg : args) {
         const char* str = arg.c_str();
         arguments[len++] = RedisModule_CreateString(ctx, str, strlen(str));
     }
-
     RedisModuleCallReply *reply = RedisModule_Call(ctx, fn_name.c_str(), "!v", arguments, len);
     RedisModule_ThreadSafeContextUnlock(ctx);
-    torch::IValue value = IValueFromRedisReply(reply);
-    RedisModule_FreeCallReply(reply);
-    RedisModule_FreeThreadSafeContext(ctx);
     for(int i= 0; i < len; i++){
-        RedisModule_FreeString(NULL, arguments[i]);
+        RedisModule_FreeString(nullptr, arguments[i]);
     }
+    torch::IValue value = IValueFromRedisReply(ctx, reply);
+    RedisModule_FreeCallReply(reply);
+    RedisModule_FreeThreadSafeContext(ctx);
     return value;
 }
 
-torch::List<torch::IValue> asList(torch::IValue v) {
+torch::List<torch::IValue> asList(const torch::IValue &v) {
     return v.toList();
 }
+
+std::vector<torch::Tensor> modelExecute(const std::string& model_key, const std::vector<torch::Tensor> &inputs, int64_t num_outputs) {
+    RedisModuleCtx* ctx = RedisModule_GetThreadSafeContext(nullptr);
+    RedisModule_ThreadSafeContextLock(ctx);
+
+    // Prepare for getting model from key space.
+    const char* model_key_str = model_key.c_str();
+    RedisModuleString *model_key_rs = RedisModule_CreateString(ctx, model_key_str,
+      strlen(model_key_str));
+    RAI_Error *err;
+    RedisAI_InitError(&err);
+    RAI_Model *model = nullptr;
+    RAI_ModelRunCtx *model_run_ctx = nullptr;
+    std::vector<torch::Tensor> outputs;
+
+    int status = RedisAI_GetModelFromKeyspace(ctx, model_key_rs, &model,
+      REDISMODULE_READ, err);
+    RedisModule_FreeString(nullptr, model_key_rs);
+    if (status != REDISMODULE_OK) {
+        RedisModule_ThreadSafeContextUnlock(ctx);
+        RedisModule_FreeThreadSafeContext(ctx);
+        goto finish;
+    }
+
+    // Create model run ctx, store the input tensors and output placeholders in it.
+    model_run_ctx = RedisAI_ModelRunCtxCreate(model);
+    RedisModule_ThreadSafeContextUnlock(ctx);
+    RedisModule_FreeThreadSafeContext(ctx);
+    for (auto &input : inputs) {
+        DLManagedTensor *dl_tensor = torchTensorPtrToManagedDLPack(&input);
+        RAI_Tensor *tensor = RedisAI_TensorCreateFromDLTensor(dl_tensor);
```

Comment:
why the deep clone and not the shallow clone?

Reply:
What do you mean? `RedisAI_TensorCreateFromDLTensor` does not perform deep copy, it takes ownership on the `dl_tensor`, and `torchTensorPtrToManagedDLPack` takes ownership on the `input` torch tensor, no? So it isn't a deep copy if I understand correct...

Reply:
wrong place of comment